gdal/frmts/netcdf/netcdfvirtual.cpp

```diff
@@ -0,0 +1,330 @@
+/******************************************************************************
+ *
+ * Project:  netCDF read/write Driver
+ * Purpose:  GDAL bindings over netCDF library.
+ * Author:   Winor Chen <wchen329 at wisc.edu>
+ *
+ ******************************************************************************
+ * Copyright (c) 2019, Winor Chen <wchen329 at wisc.edu>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+#include "netcdfvirtual.h"
+
+// netCDF Virtual
+// Provides a layer of "virtual ncID"
+// that can be mapped to a real netCDF ID
+namespace nccfdriver
+{
+    void netCDFVDimension::invalidate()
+    {
+        this->valid = false;
+        real_dim_name.clear();
+    }
+
+    netCDFVVariable::netCDFVVariable(const char * name, nc_type xtype, int ndims, const int* dimidsp, int varid) :
+        real_var_name(name),
+        ntype(xtype),
+        v_vid(varid),
+        ndimc(ndims),
+        dimid(new int[ndims])
+    {
+        for(int c = 0; c < ndims; c++)
+        {
+            dimid.get()[c] = dimidsp[c];
+        }
+    }
+
+    void netCDFVVariable::invalidate()
+    {
+        this->valid = false;
+        real_var_name.clear();
+        attribs.clear();
+    }
+
+    int netCDFVID::nc_def_vdim(const char * name, size_t len)
+    {
+        int dimID = dimTicket;
+
+        // Check if name is already defined
+        if(nameDimTable.count(std::string(name)) > 0)
+        {
+            throw SG_Exception_DupName(name, "virtual dimension collection");
+        }
+
+        // Add to lookup tables
+                dimList.push_back(netCDFVDimension(name, len, dimTicket++));
```

Comment:
indentation.

Reply:
I'd prefer dimTicket++ to be done in a separate statement

Reply:
done